### DISCUSSION TITLE
Q. Dynamic provider creation - followup

### DISCUSSION THREAD

**Discussion by adeturner** [nextauthjs/next-auth#6278] [Help]
### Question 💬

Apologies if I'm being a newbie. Further to my question #5892 unfortunately I'm still struggling to get it working. To recap, I'm trying to dynamically create a provider on signIn according to email domain. As directed I followed the [docs](https://next-auth.js.org/configuration/initialization#advanced-initialization) and instrumented my code to workout what was being passed through. 

There are not many ways to provide info through to ...[nextauth].ts from signIn.  For example when I invoke with orgdomain as a currently unknown provider:

```javascript
      signIn(orgdomain, undefined, { domain: orgdomain })
```

...[nextauth].ts auth(req, res) receives

```txt
req.method=GET req.url=/api/auth/providers req.query.nextauth=providers res.status=200
callbackRedirect url=http://localhost:3000/auth/login baseUrl=http://localhost:3000
req.method=GET req.url=/api/auth/signin?callbackUrl=http%3A%2F%2Flocalhost%3A3000%2Fauth%2Flogin req.query.nextauth=signin res.status=200
callbackRedirect url=http://localhost:3000/auth/login baseUrl=http://localhost:3000
req.method=GET req.url=/api/auth/session req.query.nextauth=session res.status=200
callbackRedirect url=http://localhost:3000/auth/login baseUrl=http://localhost:3000
req.method=GET req.url=/api/auth/session req.query.nextauth=session res.status=200
callbackRedirect url=http://localhost:3000/auth/login baseUrl=http://localhost:3000
```

I dumped the whole request object but I cannot see my provided domain anywhere.  
Please can you advise how to pass a parameter through to the place that I want to set the provider

### How to reproduce ☕️

Login submission

```javascript
    const onSubmit = (data: any) => {
      let orgdomain = getEmailDomain(data.email, true, false)
      signIn(orgdomain, undefined, { domain: orgdomain })
    }
```

[...nextauth].ts:

```javascript
export default async function auth(req: NextApiRequest, res: NextApiResponse) {
	domain = SomehowGetMyDomain(req)
	return await NextAuth(req, res, getAuthOptions(domain))
}
```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by kristiansamuelsson** [comment]
Data sent in the additional parameters are available in the request query. Note that these will also be sent to the provider.

In your example, you should be able to find it in ```req.query.domain```

https://next-auth.js.org/getting-started/client#additional-parameters

You will probably run into the same problem I did, where the callback from the auth provider does not contain any information you can use to recreate the dynamically created provider. What I did was set a cookie on signin initialization, and get the value when returning from the provider.

Another problem I faced was that using the ```signIn(...)``` function from the client with a provider that does not exists (the api/auth/providers/ endpoint does not return it) will cause the client to show an error and not initialize login. To get around that, I am always populating the list of providers with one provider with the same id as the dynamically created ones.

Here is some of my code, using login_hint instead of your "domain" paramater.

```typescript
const signinUrlPattern = '/api/auth/signin/';
const callbackUrlPattern = '/api/auth/callback/'

const PROVIDER_COOKIE_NAME = "user.provider";

export default async function auth(req: NextApiRequest, res: NextApiResponse) {
  const url = req.url as string;
  var provider = null;
  if (req.method === "POST" && url.startsWith(signinUrlPattern)) {
    const providerName = lookupProvider(req.query.login_hint as string);
    setCookie(res, PROVIDER_COOKIE_NAME, providerName);
    provider = getProvider(providerName);
  }
  else if (url.startsWith(callbackUrlPattern)) {
    const providerName = req.cookies[PROVIDER_COOKIE_NAME];
    provider = getProvider(providerName);
  }
  
  provider ??= SomeProvider({
    clientId: 'default-id',
    clientSecret: 'default-secret',
    issuer: 'https://example.com'
  });

  const providers = [provider];

  return await NextAuth(req, res, {
    providers
  })
  
//TODO: implement lookupProvider, getProvider, setCookie
}
```
